unittests/Dialect/SPIRV/Runtime/RuntimeTests.cpp

```diff
@@ -0,0 +1,235 @@
+//===- RuntimeTest.cpp - SPIR-V Runtime Tests -----------------------------===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file is for testing the Vulkan runtime API which takes in a spirv::ModuleOp,
+// a bunch of resourses and number of work groups.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SPIRV/SPIRVBinaryUtils.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Module.h"
+#include "mlir/Parser.h"
+#include "mlir/Support/FileUtilities.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+#include "gmock/gmock.h"
+#include <string>
+#include <random>
+
+using namespace mlir;
+using namespace llvm;
+
+using DescriptorSetIndex = uint32_t;
+using BindingIndex = uint32_t;
+
+// Struct containing information regarding to a host memory buffer.
+struct VulkanHostMemoryBuffer {
+  void *ptr{nullptr};
+  uint64_t size{0};
+};
+
+// Struct containing the number of local workgroups to dispatch for each
+// dimension.
+struct NumWorkGroups {
+  uint32_t x{1};
+  uint32_t y{1};
+  uint32_t z{1};
+};
+
+// This is a temporary function and will be removed in the future.
```

Comment:
Can you clarify the temporary aspect: there is a layering violation right now where you're poking at the implementation of a tools.